1단계 - EntityPersister 구현

catsbi opened the pull request:
안녕하세요. 리뷰어님, 0단계 이후 빠르게 진행하고싶었는데, 사실 감이 잘 안와서 고민을 많이 하다보니 PR이 좀 늦어졌네요. 

EntityPersister를 만들어 EntityManager의 책임을 상당부분 위임하고, PersistenceContext를 만들어 1차 캐싱 작업을 진행했습니다. 

그럼 리뷰 잘 부탁드립니다 😄

lhs0610 committed:
refactor: reformat code

lhs0610 committed:
feat: EntityPersister 인터페이스 및 구현체 구현

lhs0610 committed:
test: DefaultPersistenceContextTest 테스트 케이스 작성

lhs0610 committed:
feat: 엔티티 저장시 식별자(primary key)반환되도록 수정 및 식별자 값 Entity에 세팅되도록 함수 추가

lhs0610 committed:
feat: 병합 로직 추가

- 기존 엔티티에 내용 복사 로직 추가

lhs0610 committed:
feat: EntityPersister 구현 및 EntityManager 책임 분리및 위임

lhs0610 committed:
feat: reformatting code

jongmin4943 commented on src/main/java/persistence/sql/config/PersistenceConfig.java:
```diff
@@ -56,7 +61,15 @@ private SortedSet<QueryConstraintSupplier> constraintQuerySuppliers() {
+
+    private PersistenceContext persistenceContext() {
+        return new DefaultPersistenceContext();
     }
```
Config 를 통해 관리하고 계시군요 👍

jongmin4943 commented on src/main/java/persistence/sql/context/PersistenceContext.java:
```diff
@@ -0,0 +1,16 @@
+
+    <T, ID> T get(Class<T> entityType, ID id);
+
+    <T> Collection<T> getAll(Class<T> entityType);
```
`EntityPersister.selectAll` 은 List 를 반환하고, 여기서는 Collection 을 반환하고 있네요.

한솔님은 어떤 기준으로 반환 타입을 정하시나요? :)

catsbi replied:
저는 추상화의 수준이 높을수록 반환타입도 최대한 추상화된 객체를 사용하도록 하고 있습니다. 구현수준이 높아질수록, 구체화된 타입을 참조하도록 하구요. 좀 더 구체적으로 경계를 나눈다면, 클라이언트가 실제로 사용하는 수준에서는 퍼사드를 넘나드는 순간을 기점으로 내부는 추상화된타입을, 외부는 구체화된 타입을 사용하려고 합니다. 

그런데 여기서 EntityManager나 PersistenceContext에서 목록 정보를 가져와서 사용하는 경우 인덱스 제어를 겸하는 경우가 많기에 바로 List를 반환하도록 해도 문제가 없을 것 같네요!

jongmin4943 replied:
`추상화의 수준이 높을수록 반환타입도 최대한 추상화된 객체를 사용` 멋있네요

jongmin4943 replied:
`추상화의 수준이 높을수록 반환타입도 최대한 추상화된 객체를 사용` 멋있네요 👍

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultEntityPersister.java:
```diff
@@ -0,0 +1,142 @@
+        this.nameConverter = nameConverter;
+    }
+
+    @Override
+    public <T> Object insert(T entity, MetadataLoader<?> loader) {
+        InsertColumnValueClause clause = InsertColumnValueClause.newInstance(entity, nameConverter);
+
+        String insertQuery = QueryBuilderFactory.getInstance().buildQuery(QueryType.INSERT, loader, clause);
+        return database.executeUpdate(insertQuery);
+    }
```
깔끔하네요 💯

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultEntityPersister.java:
```diff
@@ -0,0 +1,142 @@
+        return database.executeUpdate(insertQuery);
+    }
+
+    @Override
+    public <T> boolean update(T entity, MetadataLoader<?> loader) {
+        List<Field> fields = loader.getFieldAllByPredicate(field -> !field.isAnnotationPresent(Id.class));
+
+        Clause[] clauses = new Clause[fields.size() + 1];
+        clauses[0] = WhereConditionalClause.builder()
+                .column(loader.getColumnName(loader.getPrimaryKeyField(), nameConverter))
+                .eq(Clause.toColumnValue(Clause.extractValue(loader.getPrimaryKeyField(), entity)));
+
+        for (int i = 0; i < fields.size(); i++) {
+            Field field = fields.get(i);
+            clauses[i + 1] = SetValueClause.newInstance(field, entity, loader.getColumnName(field, nameConverter));
+        }
+
+        String mergeQuery = QueryBuilderFactory.getInstance().buildQuery(QueryType.UPDATE, loader, clauses);
+        try {
+            database.executeUpdate(mergeQuery);
+            return true;
+        } catch (Exception e) {
+            e.printStackTrace();
+            return false;
+        }
+    }
```
그런데 update 는 조금 코드가 가독성이 떨어지는 것 같아요.

- `Caluse` 라는 배열을 만들때 fields.size+1 을 하는 이유를 이해하기가 쉽지 않다고 느껴져요 😢 
- clauses 의 0번째 인덱스에 `WhereConditionalClause` 를 build 하는 이유를 이해하기 쉽지 않다고 느껴져요
- 순회를 하며 clauses 에 `SetValueClause` 를 셋팅하는것도 이해하려면 알아야할 컨텍스트가 많다고 느껴져요
- 마지막에 catch 부분에서 `Exception` 이라는 상위 에러를 잡고 있어서 database 에서 에러가 나는 경우에는 문제가 생길 수 있을 것 같아요. 필요하다면 `update` 메서드의 반환타입을 void 로 하는 방향도 좋을 것 같아요 👍

개인적으로 코드 가독성을 위해 아래와 같은 접근을 해보는건 어떨까요?
- 각각 부분을 메서드로 분리해서 어떤 행위를 하는지 메서드명으로 표기해보면 좋을 것 같아요 👍 
- 해당 메서드가 만들어 졌을때 해당 행위가 `EntityPersister` 가 가진 책임과 부합하는지 보고, 아니라면 적절한 객체에게 책임을 위임해보면 좋을 것 같아요 👍 

위의 코멘트는 제 개인적인 성향이에요. 한솔님의 의견도 듣고싶네요 😃

catsbi replied:
- fields 를 기반으로 SetClause를 만드는데 WhereClause가 추가되야하기에 +1이 되었습니다. 
- 0번째 index에 넣어야 할 필수적인 이유가 있는건 아니였습니다. 그냥 whereClause를 넣어야한다는 생각으로 넣은건데, 끝에 넣어도 상관없었을 것 같네요!
- SetClause는 컬럼 하나에 매칭되는 값 정보를 담는 구분절이기에 하나 이상의 set이 필요한 경우 fields가 순회할수밖에 없을 것 같습니다. 리뷰어님이 생각하신 방안은 무엇일까요? 
- void로 변환하라는건 좋은 생각 같습니다. 수정하겠습니다. 

추가적으로 제안해주신대로 함수를 분리하여 추상화 레벨을 맞추고 네이밍을 통해 가독성을 높혀보도록 했습니다.

jongmin4943 replied:
> fields 를 기반으로 SetClause를 만드는데 WhereClause가 추가되야하기에 +1이 되었습니다.
0번째 index에 넣어야 할 필수적인 이유가 있는건 아니였습니다. 그냥 whereClause를 넣어야한다는 생각으로 넣은건데, 끝에 넣어도 상관없었을 것 같네요!
SetClause는 컬럼 하나에 매칭되는 값 정보를 담는 구분절이기에 하나 이상의 set이 필요한 경우 fields가 순회할수밖에 없을 것 같습니다. 리뷰어님이 생각하신 방안은 무엇일까요?

개인적으로 필요하다면 해당하는 책임을 가진 객체를 도출해볼 것 같아요 :)
`WhereClause` 가 추가 되어야하고, `SetClause` 를 순회해서 특정 로직을 사용해야 한다면 그 도출 된 객체에게 맡겨보면 좋지 않을까? 라고 생각되어요 😃

조금 더 제 관점을 말씀드리자면, 저는 사용하는 사람이 간단하면서 직관적이라고 느낄 수 있도록 만드려고 노력해요 :) 현재는 사용하는 사람 입장에서 조금은 알아야 할 내용이 많다고 느껴져요. 직관적으로 이해하며 사용할 수 있게 된다면 자연스레 추상화 수준도 맞지 않을까요? 👍

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultEntityPersister.java:
```diff
@@ -0,0 +1,142 @@
+        });
+    }
+
+    private <T> T mapRowResultSetToEntity(ResultSet resultSet, MetadataLoader<T> loader) {
+        try {
+            T entity = loader.getNoArgConstructor().newInstance();
+
+            ResultSetMetaData metaData = resultSet.getMetaData();
+            int columnCount = metaData.getColumnCount();
+
+            for (int i = 1; i <= columnCount; i++) {
+                Object columnValue = resultSet.getObject(i);
+
+                Field field = loader.getField(i - 1);
+                field.setAccessible(true);
+                field.set(entity, columnValue);
+            }
+
+            return entity;
+
+        } catch (ReflectiveOperationException | SQLException e) {
+            logger.severe("Failed to map row to entity");
+            throw new RuntimeException(e);
+        }
+    }
```
`Entity` 클래스 정보는 런타임 환경에서 변하지 않으니 `mapRowResultSetToEntity` 의 책임을 각 Entity 클래스에 대응하는 객체 나와서 맡아보면 어떨까요?

catsbi replied:
각 Entity(ex: Person)별로 RowMapper를 만들라는 걸까요?

jongmin4943 replied:
맞아요 :) RowMapper 생성 시 클래스 정보를 가지고 있다면 이런 느낌의 코드가 가능할 것 같아요
```java
RowMapper rowMapper = new RowMapper(Person.class);
rowMapper.mapRow(resultSet);
```

다른 관점으로는, JPA 는 DB 와의 작업이 많다보니 성능을 신경 써보는 것도 좋은 관점이라 생각해요. 

런타임 환경에서 리플렉션을 통해 클래스 정보를 가져오는 것은 인스턴스에 할당된 변수 값 같은 부분이 아니라면 매번 같은 것을 반환해요. 그럼 한번 만들어두고 재활용해서 쓰면 좋지 않을까? 라고 접근해보는거죠 👍

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -0,0 +1,80 @@
+public class DefaultPersistenceContext implements PersistenceContext {
+    private final Map<Class<?>, Map<?, Object>> context = new HashMap<>();
+
+    @Override
+    public <T, ID> T get(Class<T> entityType, ID id) {
+        if (context.containsKey(entityType) && context.get(entityType).containsKey(id)) {
+            return entityType.cast(context.get(entityType).get(id));
+        } else {
+            return null;
+        }
+    }
```
사소하지만 `객체지향생활체조 규칙 2: else 예약어를 쓰지 않는다.` 를 지켜보시는 것도 좋을 것 같아요 👍

catsbi replied:
kotlin에서 if-else 문법이 지양되지 않아서 자주 사용하다보니 습관적으로 사용한 것 같네요 🤔  수정하겠습니다.

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    @Override
+    public <T> Collection<T> getAll(Class<T> entityType) {
+        if (context.containsKey(entityType)) {
+            return context.get(entityType).values().stream()
+                    .map(entityType::cast).toList();
+        } else {
+            return Collections.emptyList();
+        }
+    }
```
`PersistenceContext` 에 getAll 이라는 메서드를 만들어주신 이유가 궁금합니다 👀

catsbi replied:
entityManager의 findAll()을 고려해서 만들었는데, 생각해보면 어느시점에 context / database에서 조회를 해야할지에 대한 조건이 명확하지가 않아보이네요. 제거하도록 하겠습니다.

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -0,0 +1,80 @@
+import java.util.Map;
+
+public class DefaultPersistenceContext implements PersistenceContext {
+    private final Map<Class<?>, Map<?, Object>> context = new HashMap<>();
```
`getAll` 이 아니었다면 현재 Map<?, Object> 로 사용되고 있는 안쪽 Map 이 바깥쪽으로 나오고, key 에 와일드카드 대신에 ID 타입을 할당할 수 있지 않았을까요?

자료구조가 조금은 복잡해서, 다른 메서드들에서 get을 한번 더 한 뒤 캐스팅해야하는 복잡도가 올라간것 같아요 😅

catsbi replied:
해당 피드백이 잘 이해가 되지 않습니다. 😢 
안쪽의 Map이 와일드카드 대신 ID 제네릭을 사용할 경우 바깥으로 나온다는게 무슨말일까요? 

객체 타입과 아이디 자체를 가지는 KeyHolder(Class<T> type, ID id)을 key로 사용하고 T를 value로 사용하는 구조를 말씀하시는걸까요? 

```java
private final Map<KeyHolder, ?> context = new HashMap<>(); 
```

이런 구조를 생각하신 걸까요?

jongmin4943 replied:
넵 맞습니다. 제가 지금 봐도 설명이 조금 이해하기 어렵네요 😅

`getAll` 메서드가 없어진다면, 굳이 Class 를 key 로 가지고 있을 필요가 있나? 싶어서요.

현재 `PersistenceContext` 가 가진 메서드들의 복잡도는 key 가 Class 인 부분 때문에 매번 반복되는 작업이 들어가기 때문이라고 느껴져요.
1. Class 로 묶인 Map<?, Object> 를 가져온다.
2. 캐스팅을 한다
3. 후 작업을 한다

말씀 해주신 형태를 만들어 가져간다면, 코드가 좀 더 간결해지지 않을까? 라고 생각해서 말씀드려봤습니다 :)

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -0,0 +1,80 @@
+        return (T) entityMap.put(id, entity);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T, ID> void merge(ID id, T entity) {
+        Map<ID, Object> entityMap = (Map<ID, Object>) context.computeIfAbsent(entity.getClass(), k -> new HashMap<>());
+
+        Object origin = entityMap.get(id);
+        if (origin == null || !isEntityChanged(entity, origin)) {
+            return;
+        }
+
+        overwriteEntity(entity, origin);
+    }
```
조금은 복잡해보이는 로직을 가졌는데 관련된 테스트가 없네요 😭

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -0,0 +1,80 @@
+        return !entity.equals(snapshot);
+    }
+
+    private <T> void overwriteEntity(T entity, Object origin) {
+        Field[] declaredFields = entity.getClass().getDeclaredFields();
+
+        for (Field field : declaredFields) {
+            field.setAccessible(true);
+            try {
+                Object value = field.get(entity);
+                field.set(origin, value);
+            } catch (IllegalAccessException e) {
+                throw new IllegalStateException("Illegal access to field: " + field.getName());
+            }
+        }
+    }
```
여기서는 만들어주신 `MetadataLoader` 가 활용 되지 않았네요. 🤔

한솔님이 만들어주신 `MetadataLoader` 의 책임은 무엇이고 언제 쓰이는걸까요?

catsbi replied:
고민만 너무 많이하다가 놓친것 같네요. 말 그대로 엔티티의 메타데이터들을 제어하는 객체로 사용하는데, Id나 Transient 애노테이션 필드 제외도 되야하는데 이를 고려하지 못한 것 같네요 🤔

jongmin4943 commented on src/main/java/persistence/sql/dml/impl/DefaultDatabase.java:
```diff
@@ -16,9 +17,19 @@ public DefaultDatabase(DatabaseServer dataSource) {
+        try (Connection connection = server.getConnection();
+             Statement statement = connection.createStatement()) {
+
+            statement.executeUpdate(query, Statement.RETURN_GENERATED_KEYS);
+
+            try (ResultSet generatedKeys = statement.getGeneratedKeys()) {
+                if (generatedKeys.next()) {
+                    return generatedKeys.getObject(1);
+                }
```
`getGeneratedKeys`의  활용 💯

jongmin4943 commented on src/main/java/persistence/sql/dml/impl/DefaultEntityManager.java:
```diff
@@ -1,53 +1,41 @@
             throw new IllegalArgumentException("Entity must not be null");
         }
-
         MetadataLoader<?> loader = new SimpleMetadataLoader<>(entity.getClass());
 
         if (!isNew(entity, loader)) {
             merge(entity);
             return;
         }
 
-        InsertColumnValueClause clause = InsertColumnValueClause.newInstance(entity, nameConverter);
-
-        String insertQuery = QueryBuilderFactory.getInstance().buildQuery(QueryType.INSERT, loader, clause);
-        database.executeUpdate(insertQuery);
+        Object id = entityPersister.insert(entity, loader);
+        updatePrimaryKeyValue(entity, id, loader);
+        persistenceContext.add(id, entity);
     }
```
persist 에서 `isNew` 를 통한 분기가 있네요.

`EntityManager` spec 에는 persist 시 이미 해당 entity 가 존재하면 Exception 을 던지라고 명시되어 있습니다.
[참고 문서](https://download.oracle.com/otn-pub/jcp/persistence-2_2-mrel-spec/JavaPersistence.pdf)
![image](https://github.com/user-attachments/assets/d8cd9503-ad5b-4214-bbd1-bf98fd3f5e6a)

`isNew` 를 통한 분기는 Spring Data Jpa 의 `SimpleJpaRepository` 가 정의하고 있습니다.
가능하시다면 `EntityManager` spec 을 확인해보시고 그에 맞춰서 구현해보려 하시는걸 권장 드립니다 👍

catsbi replied:
오 그러네요.....익숙한대로만 개발을 하다보니 놓친것 같네요. 수정하겠습니다.

jongmin4943 commented on src/main/java/persistence/sql/dml/impl/SimpleMetadataLoader.java:
```diff
@@ -25,6 +25,20 @@ public SimpleMetadataLoader(Class<T> entityType) {
         this.clazz = entityType;
     }
 
+    @Nullable
+    private static String getColumnNameByField(Field declaredField, NameConverter nameConverter) {
+        if (declaredField.isAnnotationPresent(Transient.class)) {
+            return null;
+        }
+
+        Column anno = declaredField.getAnnotation(Column.class);
+        if (anno != null && !anno.name().isBlank()) {
+            return anno.name();
+        }
+
+        return nameConverter.convert(declaredField.getName());
+    }
```
해당 메서드가 static 인 이유가 궁금합니다 👀

catsbi replied:
상태(clazz)를 사용하지 않으면서 해당 Loader를 구현하는 부분에서 공통으로 사용할 것 같아서 static으로 만들게 되었습니다

jongmin4943 replied:
저는 개인적으로 내부에서만 쓰이는 private 메서드가 public static 메서드에게 참조되지 않는다면 static 으로 만들지 않아서 여쭤 봤어요 :)

상태를 사용하지 않아서 static 으로 만든다는 관점도 신선하네요 👍

jongmin4943 commented on src/test/java/persistence/context/impl/DefaultPersistenceContextTest.java:
```diff
@@ -0,0 +1,75 @@
+        // when
+        PersonV3 actual = context.get(PersonV3.class, entity.getId());
+
+        // then
+        assertThat(actual).isNotNull();
+        assertThat(actual).isEqualTo(entity);
```
assertThat 를 여러번 호출 해야 하는 경우에 assert문이 실패하게 되면 그 후에 있는 assert 가 실행되지 않는다는 불편함이 있어요!

assetAll 이나 softAssertion 을 사용해보는시는건 어떨까요? :)
참고 링크입니다! https://zzang9ha.tistory.com/418

jongmin4943 reviewed (changes requested):
안녕하세요 한솔님!

1단계 미션 잘 진행 해주셨네요 👍
몇 가지 코드에 궁금한 부분들에 대해 코멘트 달아 보았습니다 :) 확인 후 재요청 부탁드립니다 🙇

궁금하시거나 고민 되시는 부분이 있으시다면 언제든 코멘트나 DM 주세요! 🔥

lhs0610 committed:
fix: PersistenceContext의 getAll 함수 반환타입 좀 더 구체화된 타입(List)으로 변경

lhs0610 committed:
fix: if-else pattern을 if-return방식으로 변경

lhs0610 committed:
fix: 피드백 반영

- PersistenceContext의 context 자료구조 key용 KeyHolder 객체 구현

- EntityPersister의 update, delete함수 반환타입 void로 변경

lhs0610 committed:
feat: PersistenceContext 테스트 케이스 추가

lhs0610 committed:
feat: DefaultPersistenceContext merge로직 수정

- MetadataLoader 덮어쓰기 로직 수정

lhs0610 committed:
feat: 피드백 반영

- 테스트 assertAll 사용

jongmin4943 reviewed (changes requested):
안녕하세요 한솔님!

질문 주신 부분들에 답 코멘트 달아보았습니다 :) 확인 후 재요청 부탁드립니다 🙇

궁금하시거나 고민 되시는 부분이 있으시다면 언제든 코멘트나 DM 주세요! 🔥

lhs0610 committed:
fix: 피드백 반영

- 업데이트 쿼리 구분절 객체 생성 및 테스트 케이스 수정

lhs0610 committed:
feat: RowMapper 직접 구현 및 테스트 케이스 수정

- 테스트용 Config Layer 구현

catsbi commented:
아, 브랜치를 제가 최초 2로만들어서 올리고 1로 rename하면서 바꿨더니 피드백 반영사항들을 모두 step2 브랜치로 보내서 리뷰요청할 때 정작 코드 변경사항은 없었군요 ㅠㅠ.... 

추가로 피드백주신부분 고려해서 코드 수정하려 리뷰요청드립니다!

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultEntityPersister.java:
```diff
@@ -0,0 +1,112 @@
+        return database.executeUpdate(insertQuery);
+    }
+
+    @Override
+    public <T> void update(T entity, MetadataLoader<?> loader) {
+        UpdateQueryClauses updateQueryClauses = UpdateQueryClauses.builder(nameConverter)
+                .where(entity, loader)
+                .setColumnValues(entity, loader)
+                .build();
+
+        String mergeQuery = QueryBuilderFactory.getInstance()
+                .buildQuery(QueryType.UPDATE, loader, updateQueryClauses.clauseArrays());
+        database.executeUpdate(mergeQuery);
+    }
```
코드 읽기가 좋아진 것 같아요 💯

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultEntityPersister.java:
```diff
@@ -0,0 +1,112 @@
+        database.executeUpdate(mergeQuery);
+    }
+
+    @Override
+    public <T> void delete(T entity, MetadataLoader<?> loader) {
+        Field pkField = loader.getPrimaryKeyField();
+        Object extractedValue = Clause.extractValue(pkField, entity);
+        String value = Clause.toColumnValue(extractedValue);
+
+        WhereConditionalClause clause = WhereConditionalClause.builder()
+                .column(loader.getColumnName(pkField, nameConverter))
+                .eq(value);
+
+        String removeQuery = QueryBuilderFactory.getInstance().buildQuery(QueryType.DELETE, loader, clause);
+        database.executeUpdate(removeQuery);
+    }
```
이건 반영안하셔도 되지만, 다른 메서드들과의 추상화수준을 맞추기위해 `DeleteQueryClauses` 와 같은 객체가 나와도 좋을 것 같네요 :)

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultEntityPersister.java:
```diff
@@ -0,0 +1,112 @@
+
+        String selectQuery = QueryBuilderFactory.getInstance().buildQuery(QueryType.SELECT, loader, clause);
+
+        RowMapper<T> rowMapper = rowMapperFactory.getRowMapper(entityType);
+        if (rowMapper == null) {
+            throw new IllegalStateException("RowMapper not found for entity type: " + entityType);
+        }
```
`rowMapperFactory.getRowMapper` 가 null 을 리턴하고 사용하는 쪽에서 매번 체크하는것보단, 스스로 null 을 체크하는 방식은 어떠신가요? :)

catsbi replied:
최초에는 null인경우 기존에 만들어뒀던 reflection을 이용한 row mapping작업을 수행하려고 했다가, 삭제하게되면서 throw Exception으로 바뀌었는데, 수정하겠습니다

jongmin4943 commented on src/main/java/sample/application/PersonRowMapper.java:
```diff
@@ -0,0 +1,21 @@
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class PersonRowMapper implements RowMapper<PersonV3> {
+    @Override
+    public PersonV3 mapRow(ResultSet resultSet) throws SQLException {
```
Person 용 RowMapper 를 만들어주셨네요 :) 다음 단계 진행하면서 범용적으로 사용할 수 있게 되면 좋을 것 같아요 👍

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class DefaultPersistenceContext implements PersistenceContext {
+    private final Map<KeyHolder, Object> context = new HashMap<>();
```
피드백 반영해주셨네요 🙇  코드가 좀 더 간단해진 것 같아요 👍

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -0,0 +1,79 @@
+        return null;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T, ID> T add(ID id, T entity) {
+        KeyHolder key = new KeyHolder(entity.getClass(), id);
+        if (context.containsKey(key)) {
+            return (T) context.get(key);
+        }
+
+        return (T) context.put(key, entity);
+    }
```
add 에 반환 값이 있네요. 
그런데 return 값이 해당 key 가 존재하면 기존 것을 반환하고, 없으면 항상 null 을 반환 하는데 의도 하신걸까요?

add 에 반환 값이 딱히 사용되지 않는 것 같은데 꼭 필요할까 고민이 드네요 🤔

catsbi replied:
database에 저장한다면 generated key를 새로 발급받아서 사용해야 하니 필요하지만 persistence context에선 별로 필요 없을것 같네요 🤔  제거하겠습니다.

jongmin4943 commented on src/main/java/persistence/sql/context/impl/DefaultPersistenceContext.java:
```diff
@@ -0,0 +1,79 @@
+        context.remove(key);
+    }
+
+    private boolean isEntityChanged(Object entity, Object snapshot) {
+        return !entity.equals(snapshot);
+    }
```
DirtyChecking 관련해서는 3단계 때 더 자세히 들어가보시지요 :)

jongmin4943 commented on src/main/java/persistence/sql/dml/impl/DefaultEntityManager.java:
```diff
@@ -1,53 +1,41 @@
+        persistenceContext.add(id, entity);
     }
 
     private <T> boolean isNew(Object entity, MetadataLoader<T> loader) {
```
diff 에 잡히지 않지만, `IllegalAccessException` 가 catch 되었을때 return false 를 하는게 맞을지 고민이 드네요 🤔

catsbi replied:
흠 우선 로깅 및 예외 번역하여 비검사 예외로 throw하는 방식으로 수정해봤습니다.

jongmin4943 commented on src/test/java/persistence/sql/dml/impl/DefaultEntityManagerTest.java:
```diff
@@ -1,62 +1,67 @@
-        assertThat(actual.getIndex()).isNull();
+        TestPerson actual = entityManager.find(TestPerson.class, 1L);
+
+        assertAll(
+                () -> assertThat(actual).isNotNull(),
+                () -> assertThat(actual.getName()).isEqualTo("catsbi"),
+                () -> assertThat(actual.getAge()).isEqualTo(55),
+                () -> assertThat(actual.getEmail()).isEqualTo("catsbi@naver.com"),
+                () -> assertThat(actual.getIndex()).isEqualTo(123)
+        );
```
피드백 반영 감사합니다 👍

jongmin4943 commented on src/main/java/persistence/sql/dml/EntityManager.java:
```diff
@@ -5,34 +5,41 @@
      */
     <T> void persist(T entity);
 
     /**
      * 엔티티를 수정한다.
+     *
      * @param entity 수정할 엔티티
      */
     <T> void merge(T entity);
```
`EntityManager.merge` spec 에서는 T 를 반환하고 있다는 점 참고 부탁드려요 👍
![image](https://github.com/user-attachments/assets/49f47fc4-1039-4137-a4fc-48a98f4a195d)

jongmin4943 reviewed (approved):
안녕하세요 한솔님!

피드백에 대해 고민 해주시고 빠르게 반영 해주셨네요 💯
사소한 코멘트 남겨두었으니 다음 단계 진행하시면서 반영해주세요. 🙇

그럼 이번 단계는 merge 하겠습니다 🚀
궁금한 사항은 언제든 추가 코멘트 또는 DM 주세요! 😃
